File: fortran/src/source/rk4.f
C ===== BEGIN FILE =============================================================
C **                                                                          **
C **  COPYRIGHT (C) 2007, STEPHEN W. SOLIDAY                                  **
C **                      STEPHEN@SOLIDAY.COM                                 **
C **                      HTTP://WWW.SOLIDAY.COM/STEPHEN                      **
C **                                                                          **
C **  THIS PROGRAM IS FREE SOFTWARE: YOU CAN REDISTRIBUTE IT AND/OR MODIFY    **
C **  IT UNDER THE TERMS OF THE GNU GENERAL PUBLIC LICENSE AS PUBLISHED BY    **
C **  THE FREE SOFTWARE FOUNDATION, EITHER VERSION 3 OF THE LICENSE, OR       **
C **  (AT YOUR OPTION) ANY LATER VERSION                                      **
C **                                                                          **
C **  THIS PROGRAM IS DISTRIBUTED IN THE HOPE THAT IT WILL BE USEFUL, BUT     **
C **  WITHOUT ANY WARRANTY; WITHOUT EVEN THE IMPLIED WARRANTY OF              **
C **  MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. SEE THE            **
C **  GNU GENERAL PUBLIC LICENSE FOR MORE DETAILS.                            **
C **                                                                          **
C **  YOU SHOULD HAVE RECEIVED A COPY OF THE GNU GENERAL PUBLIC LICENSE       **
C **  ALONG WITH THIS PROGRAM. IF NOT, SEE <HTTP://WWW.GNU.ORG/LICENSES/>.    **
C **                                                                          **
C **  ----- MODIFICATION HISTORY -------------------------------------------- **
C **                                                                          **
C **  AUTHOR STEPHEN W. SOLIDAY                                               **
C **  DATE   2002-02-18                                                       **
C **                                                                          **
C **  PROVIDES THE IMPLEMENTATION FOR THE FOURTH ORDER RUNGE-KUTTA NUMERICAL  **
C **  INTEGRATOR.                                                             **
C **                                                                          **
C ==============================================================================

C----+==================================================================--------
      BLOCK DATA RK4BD
C----+==================================================================--------
      DOUBLE PRECISION A,B,C,D,W
      DIMENSION A(64),B(64),C(64),D(64),W(64)
      COMMON /RK4PARS/A,B,C,D,W
C     ------------------------------------------------------------------
      END

C----+==================================================================--------
C     FOURTH ORDER RUNGE-KUTTA.
C     PROVIDES THE IMPLEMENTATION FOR A FOURTH ORDER RUNGE-KUTTA NUMERICAL
C     INTEGRATOR WITH UNIFORM STEP SIZES.
C    \PARAM Q REAL VECTOR CONTAINING THE STATE.
C    \PARAM N NUMBER OF ELEMENTS IN THE STATE VECTOR.
C    \PARAM T0 INITIAL TIME.
C    \PARAM T1 FINAL TIME.
C    \PARAM STEP NUMBER OF STEPS BETWEEN CURRENT TIME A T0 AND FINAL TIME A T1.
C    \PARAM P VECTOR CONTAINING FIXED PARAMETERS.
C    \RETURN NEW TIME (\A T1).
C----+==================================================================--------
      FUNCTION RK4(Q,N,T0,T1,STEP,P)
      DOUBLE PRECISION RK4
C----+==================================================================--------
      DOUBLE PRECISION Q(*)
      INTEGER N
      DOUBLE PRECISION T0,T1
      INTEGER STEP
      DOUBLE PRECISION P(*)
C     ------------------------------------------------------------------
      DOUBLE PRECISION A,B,C,D,W
      DIMENSION A(64),B(64),C(64),D(64),W(64)
      COMMON /RK4PARS/A,B,C,D,W
C     ------------------------------------------------------------------
      DOUBLE PRECISION H,H2,T
      INTEGER J,K,R, CHECK
C     ------------------------------------------------------------------
      H = (T1-T0)/DBLE(STEP)
      T=T0
      H2=H/2.0

      DO 10 K=1,STEP
         R = CHECK(Q,T,P)
         IF(0.LT.R) THEN
            WRITE(0,*) '2 CHECK FAILED'         
            GOTO 20
         END IF
C        --------------------------------------
         DO 1 J=1,N
            W(J)=Q(J)
 1       CONTINUE
         CALL DIFEQ(A,W,T,P)
C        --------------------------------------
         DO 2 J=1,N
            W(J)=Q(J)+(A(J)*H2)
 2       CONTINUE
         CALL DIFEQ(B,W,T+H2,P)
C        --------------------------------------
         DO 3 J=1,N
            W(J)=Q(J)+(B(J)*H2)
 3       CONTINUE
         CALL DIFEQ(C,W,T+H2,P)
C        --------------------------------------
         DO 4 J=1,N
            W(J)=Q(J)+(C(J)*H)
 4       CONTINUE
         CALL DIFEQ(D,W,T+H,P)
C        --------------------------------------
         DO 5 J=1,N
            Q(J)= Q(J)+(H*(A(J)+2.0*(B(J)+C(J))+D(J))/6.0)
 5          CONTINUE
C        --------------------------------------
         T=T+H
C        --------------------------------------
 10   CONTINUE
 20   CONTINUE
      RK4=T
C     ------------------------------------------------------------------
      RETURN
      END

C =============================================================== END FILE =====

File: c/src/source/rk4.c
/* ===== BEGIN FILE ========================================================================= */
/* **                              P R O P R I E T A R Y                                   ** */
/* ========================================================================================== */
/* **                                                                                      ** */
/* **  Copyright (c) 2006, Stephen W. Soliday                                              ** */
/* **                      stephen@soliday.com                                             ** */
/* **                      http://www.soliday.com/stephen                                  ** */
/* **                                                                                      ** */
/* **  This file, and the associated algorithms, are not free software; you may not        ** */
/* **  redistribute them and/or modify them. These algorithms were developed and           ** */
/* **  implemented for the purpose of an internal assessment and have, as yet, not been    ** */
/* **  publicly distributed. Development of these algorithms have been at the sole cost    ** */
/* **  in both time and funding by their author. Until such a public release is made,      ** */
/* **  the author retains ALL RIGHTS to these algorithms. It is expected that if this      ** */
/* **  program or any of the algorithms contained herein are deemed releasable they will   ** */
/* **  be released under the GNU Public license for non-commercial use and/or with         ** */
/* **  restricted rights for government use. At that time each source file will contain    ** */
/* **  either/both the standard GPL statement/disclaimer, and/or                           ** */
/* **  the DFARS Restricted Rights Legend.                                                 ** */
/* **                                                                                      ** */
/* **  These algorithms exists at the present time WITHOUT ANY WARRANTY; without even      ** */
/* **  the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.        ** */
/* **  As you are not supposed to be in possession of this file if you use it,             ** */
/* **  you do so AT YOUR OWN RISK.                                                         ** */
/* **                                                                                      ** */
/* **  ----- Modification History -------------------------------------------------------  ** */
/* **                                                                                      ** */
/* **  Author Stephen W. Soliday                                                           ** */
/* **  Date   2006-11-24                                                                   ** */
/* **                                                                                      ** */
/**   \file rk4.c
 *    \brief Implementation.
 *     Provides the implementation for the fourth order Runge-Kutta numerical integrator.
 */
/* ========================================================================================== */

#include <rk4.h>

static double A[64]; /**< stage one state vector */
static double B[64]; /**< stage two state vector */
static double C[64]; /**< stage three state vector */
static double D[64]; /**< stage four state vector */

static double W[64]; /**< DIFEQ input vector */

/* ========================================================================================== */
/** Fourth order Runge-Kutta.
 *  Provides the implementation for a fourth order Runge-Kutta numerical integrator with 
 *  uniform step sizes.
 * \param Q real vector containing the state.
 * \param n number of elements in the state vector.
 * \param t0 initial time.
 * \param t1 final time.
 * \param step number of steps between current time \a t0 and final time \a t1.
 * \param P vector containing fixed parameters.
 * \return new time (\a t1).
 */
/* ------------------------------------------------------------------------------------------ */
double rk4( double *Q, int n, double t0, double t1, int step, double *P ) {
  /* ---------------------------------------------------------------------------------------- */

  double h = (t1 - t0) / ((double) step);

  double t = t0;

  double h2 = h/2.0;

  int k, j;

  for (k=0; k<step; k++) {

    int r = CHECK( Q, t, P );

    if (0 < r) { 
      fprintf( stderr, "Check Failed\n" );
      return t;
    }

    for (j=0; j<n; j++) { W[j] = Q[j];               } DIFEQ(A, W, t,    P);
    for (j=0; j<n; j++) { W[j] = Q[j] + (A[j] * h2); } DIFEQ(B, W, t+h2, P);
    for (j=0; j<n; j++) { W[j] = Q[j] + (B[j] * h2); } DIFEQ(C, W, t+h2, P);
    for (j=0; j<n; j++) { W[j] = Q[j] + (C[j] * h);  } DIFEQ(D, W, t+h,  P);

    for (j=0; j<n; j++) { Q[j] += (h*(A[j] + 2.0*(B[j] + C[j]) + D[j])/6.0); }

    t += h;
  }

  return t;
}

/* ========================================================================================== */
/* **                              P R O P R I E T A R Y                                   ** */
/* =========================================================================== END FILE ===== */
